src/algebra/star/self_adjoint.lean

```diff
@@ -0,0 +1,114 @@
+/-
+Copyright (c) 2021 Frédéric Dupuis. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Frédéric Dupuis
+-/
+
+import algebra.star.basic
+import group_theory.subgroup.basic
+
+/-!
+# Self-adjoint elements of a star additive group
+
+This file defines `self_adjoint R`, where `R` is a star additive monoid, as the additive subgroup
+containing the elements that satisfy `star x = x`. This includes, for instance, Hermitian
+operators on Hilbert spaces.
+
+## TODO
+
+* If `R` is a `star_module R₂ R`, put a module structure on `self_adjoint R`. This would naturally
+be a `module (self_adjoint R₂) (self_adjoint R)`, but doing this literally would be undesirable
+since in the main case of interest (`R₂ = ℂ`) we want `module ℝ (self_adjoint R)` and not
+`module (self_adjoint ℂ) (self_adjoint R)`.
```

Comment:
Can you mention some possibilites for working around this?  Here's a method that occurs to me:

first step:
```lean
variables (𝕜 : Type*) [ring 𝕜]
variables (E : Type*) [add_comm_group E] [module 𝕜 E] [star_add_monoid E]

class star_invariant_scalars : Prop :=
(invariant : ∀ x : 𝕜, ∀ v : E, star (x • v) = x • star v)

instance [star_invariant_scalars 𝕜 E] : module 𝕜 (self_adjoint E) := sorry
```

second step:
```lean
variables (𝕜 : Type*) [comm_ring 𝕜]
variables (𝕂 : Type*) [ring 𝕂] [star_ring 𝕂] [algebra 𝕜 𝕂]
variables (E : Type*) [add_comm_group E] [module 𝕜 E] [star_add_monoid E]
  [module 𝕂 E] [star_module 𝕂 E] [is_scalar_tower 𝕜 𝕂 E]

instance [star_invariant_scalars 𝕜 𝕂] : star_invariant_scalars 𝕜 E := sorry
```

Reply:
(Here the `𝕜` would be `ℝ`, the `𝕂` would be `ℂ`, and then after proving `star_invariant_scalars ℝ ℂ` you'd be good to go with `module ℝ (self_adjoint E)` for any star-module `E` over `ℂ`.)

Reply:
I was thinking of a similar trick: define `[has_trivial_star E]` which states that the star operation is trivial. Then, if we have a `[module 𝕜 E]`, where `𝕜` has a trivial star, we automatically get a `[module 𝕜 (self_adjoints E)]`. (Here of course `ℝ` would be the type with a trivial star.)

Reply:
I now briefly outlined those two ideas in the TODO.

Reply:
Yup, makes sense.  In your idea you would also need to assume `[star_module 𝕜 E]`, right?  Anyway I think this is enough information to guide a future user, thank you for adding the summary!

Reply:
Probably -- I haven't fully thought it through yet!